lib/http_v4_signature.c

```diff
@@ -0,0 +1,310 @@
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+#include "parsedate.h"
+
+#include <time.h>
+
+static CURLcode hmac_sha256(const unsigned char *key, unsigned int keylen,
+                            const unsigned char *data, unsigned int datalen,
+                            unsigned char *output)
+{
+  struct HMAC_context *ctxt = Curl_HMAC_init(Curl_HMAC_SHA256, key, keylen);
+
+  if(!ctxt)
+    return CURLE_OUT_OF_MEMORY;
+
+  /* Update the digest with the given challenge */
+  Curl_HMAC_update(ctxt, data, datalen);
+
+  /* Finalise the digest */
+  Curl_HMAC_final(ctxt, output);
+
+  return CURLE_OK;
+}
+
+#define HMAC_SHA256(k, kl, d, dl, o)                  \
+  if(hmac_sha256(k, kl, d, dl, o) != CURLE_OK) {      \
+    ret = CURLE_OUT_OF_MEMORY;                        \
+    goto free_all;                                    \
+  }
+
+#define PROVIDER_MAX_L 16
+
+CURLcode Curl_output_v4_signature(struct connectdata *conn, bool proxy)
+{
+  CURLcode ret = CURLE_OK;
+  char sk[45]; /* secret key is 40 chat long + 'OSC' + \0 */
+  struct Curl_easy *data = conn->data;
+  const char *customrequest = data->set.str[STRING_CUSTOMREQUEST];
+  const char *surl = strstr(data->set.str[STRING_SET_URL], "://") + 3;
+  char *host = NULL;
+  struct tm info;
+  time_t rawtime;
+  const char *provider = data->set.str[STRING_V4_SIGNATURE];
+  char low_provider0[PROVIDER_MAX_L + 1];
+  char low_provider[PROVIDER_MAX_L + 1];
+  char up_provider[PROVIDER_MAX_L + 1];
+  char mid_provider[PROVIDER_MAX_L + 1];
+  char *region = NULL;
+  char *uri = NULL;
+  char *query_url = NULL;
+  char *api_type = NULL;
+  char date_iso[17];
+  char date[9];
+  char date_str[64];
+  const char *post_data = data->set.postfields ?
+    data->set.postfields : "";
+  const char *content_type = Curl_checkheaders(conn, "Content-Type");
+  unsigned char sha_d[32];
+  char sha_str[65];
+  char *cred_scope = NULL;
+  char *signed_headers = NULL;
+  char request_type[PROVIDER_MAX_L + sizeof("4_request")];
+  char *canonical_hdr = NULL;
+  char *canonical_request = NULL;
+  char *str_to_sign = NULL;
+  unsigned char tmp_sign0[32];
+  unsigned char tmp_sign1[32];
+  char *auth = NULL;
+  char *tmp;
+  int i;
+
+  (void) proxy;
+  if(Curl_checkheaders(conn, "Authorization")) {
+    /* Authorization already present, Bailing out */
+    return CURLE_OK;
+  }
+
+  if(!customrequest)
+    customrequest = "POST";
+  if(content_type) {
+    content_type = strchr(content_type, ':') + 1;
+    /* Skip whitespace now */
+    while(isblank(*content_type))
+      ++content_type;
+  }
+
+  /* Get Parameter
+     Google and Outscale use the same OSC or GOOG,
+     but Amazon use AWS and AMZ for header arguments */
+  tmp = strchr(provider, ':');
+  if(tmp) {
+    for(i = 0; provider != tmp; ++provider, ++i) {
+      if(i > PROVIDER_MAX_L)
+        return CURLE_FAILED_INIT;
+      up_provider[i] = (char)toupper(*provider);
+      low_provider0[i] = (char)tolower(*provider);
```

Comment:
toupper and tolower cannot be used (they're locale dependent). We provide `Curl_raw_toupper()` that's safe to use.